README.md
Utilities for accessing builds of the Dart SDK.


CHANGELOG.md
## 0.3.0

* **Breaking:** Changes the name of multiple methods to match effective dart standards
  * `DartDownloads#getDownloadLink` is now `DartDownloads#createDownloadLink`
  * `DartDownloads#getVersions` is now `DartDownloads#fetchVersions`
  * `DartDownloads#getVersionPaths` is now `DartDownloads#fetchVersionPaths`
  * `DartDownloads#getVersion` is now `DartDownloads#fetchVersion`
* Update to support null safety
* Update dependencies
* Fix various lints

## 0.1.0

* Added `VersionInfo`

* `DartDownloads`
  * `getVersion` now returns `VersionInfo`
  * Added `getVersions`
  * Removed `getVersionMap`

## 0.0.1

* Just getting started.


test/sdk_builds_test.dart
library sdk_builds.test;

import 'package:pub_semver/pub_semver.dart';
import 'package:sdk_builds/sdk_builds.dart';
import 'package:test/test.dart';

late DartDownloads _dd;

void main() async {
  setUp(() {
    _dd = DartDownloads();
  });

  tearDown(() {
    _dd.close();
  });

  test('fetch version', () async {
    const revision = '31822';
    const channel = 'stable';

    final version = await _dd.fetchVersion(channel, revision);

    expect(version.version, Version.parse('1.1.1'));
  });

  test('fetch version paths', () async {
    const channel = 'stable';

    final paths = await _dd.fetchVersionPaths(channel).toList();

    expect(paths, contains('channels/stable/release/29803/'));
    expect(paths, contains('channels/stable/release/latest/'));
    expect(paths.length, greaterThanOrEqualTo(28));
  });

  test('fetchVersions', () async {
    final versions = await _dd.fetchVersions('stable');

    expect(versions.length, greaterThanOrEqualTo(27));
  });

  group('fetchVersion', () {
    test('recent git build', () async {
      const channel = 'dev';
      const revision = '1.11.0-dev.5.2';

      final content = await _dd.fetchVersion(channel, revision) as GitVersionInfo;

      expect(content.ref, '23736d3630da614c655d0569e1ba5af2021b1c61');
      expect(content.version, Version.parse('1.11.0-dev.5.2'));
      expect(content.date, DateTime(2015, 6, 11));
    });

    test('recent svn build', () async {
      const channel = 'stable';
      const revision = '44672';

      final content = await _dd.fetchVersion(channel, revision) as SvnVersionInfo;

      expect(content.revision, 44672);
      expect(content.version, Version.parse('1.9.1'));
      expect(content.date, DateTime.utc(2015, 3, 25, 3, 47));
    });

    test('old revision with old date format', () async {
      // dev/release/30039 - 0.8.10.8_r30039
      final content = await _dd.fetchVersion('dev', '30039') as SvnVersionInfo;

      expect(content.revision, 30039);
      expect(content.version, Version.parse('0.8.10-rev.8.30039'));
      expect(content.date, DateTime.utc(2013, 11, 7, 3, 48));
    });
  });

  test('downloadLink', () async {
    const channel = 'stable';
    const revision = '44672';

    final content = await _dd.createDownloadLink(
        channel, revision, 'api-docs/dart-api-docs.zip');
    expect(content.pathSegments, [
      'download',
      'storage',
      'v1',
      'b',
      'dart-archive',
      'o',
      'channels/$channel/release/$revision/api-docs/dart-api-docs.zip'
    ]);
  });
}


web/index.dart
library sdk_builds.web.index;

import 'dart:async';

import 'package:stack_trace/stack_trace.dart';

import "package:angular2/angular2.dart";
import "package:angular2/src/reflection/reflection.dart" show reflector;
import "package:angular2/src/reflection/reflection_capabilities.dart"
    show ReflectionCapabilities;

import 'package:http/browser_client.dart';
import 'package:sdk_builds/sdk_builds.dart' as sdk;

@Component(selector: "hello-app", lifecycle: const [onInit])
@View(templateUrl: 'template.html', directives: const [NgFor])
class ArchiveComponent {
  final _dd = new sdk.DartDownloads(client: new BrowserClient());

  final String channel = 'dev';

  String selectedOs = '';
  String selectedRelease = '';

  bool loading = true;

  final List<String> operatingSystems = new List<String>.unmodifiable(
      new List.from(sdk.platforms)..insert(0, 'All'));

  final List<String> releases = <String>[];

  void onInit() {
    _updateVersions();
  }

  Future _updateVersions() async {
    var items = await _dd.getVersions(channel);

    releases.add('All');
    releases.addAll(items.reversed.map((v) => v.toString()));

    // work-around for https://github.com/angular/angular/issues/2520
    // not awaiting this – want it run async
    new Timer(const Duration(seconds: 0), () {
      this.selectedRelease = items.last.toString();
      this.selectedOs = 'Mac';

      loading = false;
    });
  }
}

main() {
  Chain.capture(() {
    reflector.reflectionCapabilities = new ReflectionCapabilities();
    bootstrap(ArchiveComponent);
  }, onError: (error, Chain chain) {
    print(error);
    print(chain.terse);
  });
}


lib/src/version_info.dart
library sdk_builds.version_info;

import 'package:pub_semver/pub_semver.dart';

final _oldRevisionPostfix = new RegExp(r'(\d+\.\d+\.\d+)\.(\d+)_r(\d+)');

abstract class VersionInfo implements Comparable<VersionInfo> {
  final Version version;
  final DateTime date;
  final String channel;
  final String revisionPath;

  VersionInfo(this.version, this.date, this.channel, this.revisionPath);

  static VersionInfo parse(
      String channel, String revisionPath, Map<String, dynamic> json) {

    // Date parse magic
    var dateJson = json['date'];
    DateTime date;
    try {
      date = DateTime.parse(dateJson);
    } on FormatException {
      // dealing with weird format: 201401150424
      dateJson = '${dateJson.substring(0, 8)}T${dateJson.substring(8,12)}Z';
      date = DateTime.parse(dateJson);
    }

    // Version logic
    var jsonVersion = json['version'];

    var oldMatch = _oldRevisionPostfix.firstMatch(jsonVersion);
    if (oldMatch != null) {
      jsonVersion = "${oldMatch[1]}-rev.${oldMatch[2]}.${oldMatch[3]}";
    }

    var version = new Version.parse(jsonVersion);

    var revision = json['revision'];
    var svnRevision = int.parse(revision, onError: (_) => null);
    if (svnRevision == null) {
      // assume git!
      assert(revision is String);
      assert(revision.length == 40);

      return new GitVersionInfo(version, date, channel, revisionPath, revision);
    }

    return new SvnVersionInfo(
        version, date, channel, revisionPath, svnRevision);
  }

  String toString() => version.toString();

  int compareTo(VersionInfo vi) => this.version.compareTo(vi.version);
}

class SvnVersionInfo extends VersionInfo {
  final int revision;

  SvnVersionInfo(Version version, DateTime date, String channel,
      String revisionPath, this.revision)
      : super(version, date, channel, revisionPath);
}

class GitVersionInfo extends VersionInfo {
  final String ref;

  GitVersionInfo(Version version, DateTime date, String channel,
      String revisionPath, this.ref)
      : super(version, date, channel, revisionPath);
}


lib/src/dart_downloads.dart
library sdk_builds.dart_downloads;

import 'dart:convert';

import 'package:googleapis/storage/v1.dart' as storage;
import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;

import 'version_info.dart';

const _dartChannel = 'dart-archive';
const _flavor = 'release';

String _revisionPath(String channel, String revision,
        [List<String> extra = const []]) =>
    p.joinAll(['channels', channel, _flavor, revision, ...extra]);

class DartDownloads {
  final storage.StorageApi _api;
  final http.Client _client;

  DartDownloads._(http.Client client)
      : _client = client,
        _api = storage.StorageApi(client);

  /// If [client] is provided, it will be closed with the call to [close].
  factory DartDownloads({http.Client? client}) =>
      DartDownloads._(client ?? http.Client());

  Future<Uri> createDownloadLink(
      String channel, String revision, String path) async {
    final prefix = _revisionPath(channel, revision, [path]);

    final results = await _api.objects.list(_dartChannel, prefix: prefix);

    final items = results.items;

    if (items == null || items.isEmpty) {
      throw 'no items found for path $path';
    } else if (items.length > 1) {
      throw 'too many items for path $path';
    }

    final mediaLink = items.single.mediaLink;

    if (mediaLink == null) {
      throw 'no media link present for path $path';
    } else {
      return Uri.parse(mediaLink);
    }
  }

  Future<List<VersionInfo>> fetchVersions(String channel) async {
    final versions = await fetchVersionPaths(channel)
        .where((event) => !event.contains('latest'))
        .toList();

    final versionMaps = <VersionInfo>[];

    await Future.forEach<String>(versions, (path) async {
      try {
        final revisionString = p.basename(path);

        final ver = await fetchVersion(channel, revisionString);

        versionMaps.add(ver);
      } catch (e) {
        // TODO: some kind of log?
        print('Error with $path - $e');
      }
    });

    versionMaps.sort();

    return versionMaps;
  }

  Stream<String> fetchVersionPaths(String channel) async* {
    final prefix = p.join('channels', channel, _flavor) + '/';

    String? nextToken;

    do {
      final objects = await _api.objects.list(_dartChannel,
          prefix: prefix, pageToken: nextToken, delimiter: '/');
      nextToken = objects.nextPageToken;

      final prefixes = objects.prefixes;

      if (prefixes == null) {
        continue;
      }

      for (var item in prefixes) {
        yield item;
      }
    } while (nextToken != null);
  }

  Future<VersionInfo> fetchVersion(String channel, String revision) async {
    final media = await _fetchFile(channel, revision, 'VERSION');

    final json = await _jsonAsciiDecoder
        .bind(media.stream)
        .cast<Map<String, dynamic>>()
        .first;

    return VersionInfo.parse(channel, revision, json);
  }

  void close() => _client.close();

  Future<storage.Media> _fetchFile(
      String channel, String revision, String path) async {
    return await _api.objects.get(
        _dartChannel, _revisionPath(channel, revision, [path]),
        downloadOptions: storage.DownloadOptions.fullMedia) as storage.Media;
  }
}

late final _jsonAsciiDecoder = json.fuse(ascii).decoder;


lib/sdk_builds.dart
library sdk_builds;

import 'dart:async';
import 'dart:convert';

import 'package:googleapis/storage/v1.dart' as storage;
import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;
import 'package:pub_semver/pub_semver.dart';
import 'package:quiver/async.dart' as qa;

const _dartChannel = 'dart-archive';
const _flavor = 'release';

String _revisionPath(String channel, String revision,
    [List<String> extra = const []]) => p.joinAll(
        [['channels', channel, _flavor, revision], extra].expand(((e) => e)));

final _oldRevisionPostfix = new RegExp(r'(\d+\.\d+\.\d+)\.(\d+)_r(\d+)');

abstract class VersionInfo implements Comparable<VersionInfo> {
  final Version version;
  final DateTime date;
  final String channel;
  final String revisionPath;

  VersionInfo(this.version, this.date, this.channel, this.revisionPath);

  static VersionInfo parse(
      String channel, String revisionPath, Map<String, dynamic> json) {

    // Date parse magic
    var dateJson = json['date'];
    DateTime date;
    try {
      date = DateTime.parse(dateJson);
    } on FormatException {
      // dealing with weird format: 201401150424
      dateJson = '${dateJson.substring(0, 8)}T${dateJson.substring(8,12)}Z';
      date = DateTime.parse(dateJson);
    }

    // Version logic
    var jsonVersion = json['version'];

    var oldMatch = _oldRevisionPostfix.firstMatch(jsonVersion);
    if (oldMatch != null) {
      jsonVersion = "${oldMatch[1]}-rev.${oldMatch[2]}.${oldMatch[3]}";
    }

    var version = new Version.parse(jsonVersion);

    var revision = json['revision'];
    var svnRevision = int.parse(revision, onError: (_) => null);
    if (svnRevision == null) {
      // assume git!
      assert(revision is String);
      assert(revision.length == 40);

      return new GitVersionInfo(version, date, channel, revisionPath, revision);
    }

    return new SvnVersionInfo(
        version, date, channel, revisionPath, svnRevision);
  }

  String toString() => version.toString();

  int compareTo(VersionInfo vi) => this.version.compareTo(vi.version);
}

class SvnVersionInfo extends VersionInfo {
  final int revision;

  SvnVersionInfo(Version version, DateTime date, String channel,
      String revisionPath, this.revision)
      : super(version, date, channel, revisionPath);
}

class GitVersionInfo extends VersionInfo {
  final String ref;

  GitVersionInfo(Version version, DateTime date, String channel,
      String revisionPath, this.ref)
      : super(version, date, channel, revisionPath);
}

class DartDownloads {
  final storage.StorageApi _api;
  final http.Client _client;

  DartDownloads._(http.Client client)
      : this._client = client,
        this._api = new storage.StorageApi(client);

  /// If [client] is provided, it will be closed with the call to [close].
  factory DartDownloads({http.Client client}) {
    if (client == null) {
      client = new http.Client();
    }
    return new DartDownloads._(client);
  }

  Future<Uri> getDownloadLink(
      String channel, String revision, String path) async {
    var prefix = _revisionPath(channel, revision, [path]);

    storage.Objects results =
        await _api.objects.list(_dartChannel, prefix: prefix);

    if (results.items == null || results.items.isEmpty) {
      throw 'no items found for path $path';
    } else if (results.items.length > 1) {
      throw 'too many items for path $path';
    }

    storage.Object obj = results.items.single;

    return Uri.parse(obj.mediaLink);
  }

  Future<String> getHash256(
      String channel, String revision, String download) async {
    var media = await _api.objects.get(
        _dartChannel, _revisionPath(channel, revision, ['$download.sha256sum']),
        downloadOptions: storage.DownloadOptions.FullMedia);

    var hashLine = await ASCII.decodeStream(media.stream);
    return new RegExp('[0-9a-fA-F]*').stringMatch(hashLine);
  }

  Future<List<VersionInfo>> getVersions(String channel) async {
    var versions = await getVersionPaths(channel).toList();

    versions.removeWhere((e) => e.contains('latest'));

    var versionMaps = <VersionInfo>[];

    await qa.forEachAsync(versions, (path) async {
      try {
        var revisionString = p.basename(path);

        var ver = await getVersion(channel, revisionString);

        versionMaps.add(ver);
      } catch (e) {
        // TODO: some kind of log?
        print("Error with $path - $e");
        return;
      }
    }, maxTasks: 6);

    versionMaps.sort();

    return versionMaps;
  }

  Stream<String> getVersionPaths(String channel) async* {
    var prefix = p.join('channels', channel, _flavor) + '/';

    var nextToken = null;

    do {
      storage.Objects objects = await _api.objects.list(_dartChannel,
          prefix: prefix, pageToken: nextToken, delimiter: '/');
      nextToken = objects.nextPageToken;

      if (objects.prefixes == null) {
        continue;
      }

      for (var item in objects.prefixes) {
        yield item;
      }
    } while (nextToken != null);
  }

  Future<VersionInfo> getVersion(String channel, String revision) async {
    var media = await _getFile(channel, revision, 'VERSION');

    var json = await _jsonAsciiDecoder.bind(media.stream).first;

    return VersionInfo.parse(channel, revision, json);
  }

  void close() => _client.close();

  Future<storage.Media> _getFile(
      String channel, String revision, String path) async {
    return await _api.objects.get(
        _dartChannel, _revisionPath(channel, revision, [path]),
        downloadOptions: storage.DownloadOptions.FullMedia);
  }
}

final _jsonAsciiDecoder = JSON.fuse(ASCII).decoder;
